### PR TITLE
Windows linux translation

### PR DESCRIPTION
**Pull request by Ryan-DowlingSoka**
This is my first submission in an attempt to update the Image Magick docs to help out windows users by doing auto conversion between linux cli syntax and windows.

It is heavily inspired by https://imagemagick.org/Usage/windows/ which contains a script to do just this.

I have not 1 to 1 implemented the script due to some complexities I'm not sure were fully covered, such as inline comments, but did an additional fix of ensuring spaces on either side of parens which is a requirement of dos.

Here is a rundown:

- The logic is implemented in **cli-translate.js** it works by finding any `<pre>` tags with the "cli" class and then copies the contents 2 times.
- The first copy is translated to the windows syntax, then the second copy is translated into the windows bat file syntax. (Double percentages for example is required for bat files, but not for the dos or powershell commandlines.)
- The conversions are done with regex and replaces, similar in design to the script at Usage/windows.
- Once both version are translated if the batch version is the same as the linux original version then I early out, but if it is different then they are added to the DOM and so is a combo-box button to the bottom right of the box.
- When this button switches, _all_ code boxes are swapped at the same time. I chose to do it this way since it seemed more likely that the general use case will be a user setting their desired cli language once for a page instead of wanting to see one command box one way and another a different way.
  - It would be possible to re-engineer this to be a box by box setting, but I think that is a worse user experience.
  - It would also be possible to store the language into the URL or into a cookie, whichever is preferred, but I personally don't think it is that necessary.
- The javascript is minified and echo'd into the magick.js as a local addition.
- No external libraries are used.
- I added the cli tag to all matches of: `<pre "*"><samp>magick` 
  - It is possible some were missed, especially any that were written a different way, perhaps in a `<code>` box.
  - Any interpreter and results implementations `$ magick ...` were missed, but this was to avoid mangling their return results.
  - Any linux bash files were missed, as I did not convert their headers and did not implement support for inline comments.

Here is a gif showing what the system looks like in practice.
imagemagic_docs_2 gif

I did spot check several image magick commands using both the dos commandline (windows 10) and .bat files

I had to fix the very first command which had an errant `")` at the end.

If this goes well I would like to bring this logic (modified of course) to the usage examples, which will require a bit more effort but I think be very valuable.

### CODE REVIEW

File: assets/magick-js.php [hunk 1]
```diff
@@ -6,5 +6,7 @@
   header('Expires: ' . gmdate('D, d M Y H:i:s',time()+604800) . ' GMT');
   header("Content-type: text/javascript; charset=utf-8");
   readfile('bootstrap.bundle.min.js');
+  echo(PHP_EOL);
+  readfile('cli-translate.min.js');
```
Comment (Ryan-DowlingSoka) [inline assets/magick-js.php:10]: Here is where I add the cli-translate.min.js
I had to add the PHP_EOL because the bootstrap.bundle.min.js ends with a comment line and not a newline.

Comment (Ryan-DowlingSoka) [inline assets/magick-js.php:10]: Here is where I add the cli-translate.min.js
I had to add the PHP_EOL because the bootstrap.bundle.min.js ends with a comment line and not a newline.

File: source/cli-translate.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
+var imagemagick_cli_lists = null;
```
Comment (Ryan-DowlingSoka) [inline source/cli-translate.js:1]: the generated lists are done once. this _will not work_ if the site changes to load content dynamically since it will be too late. This could easily be changed if there is a need.

File: source/cli-translate.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+function addSelectOptions(node)
+{
+    var selectCliLanguageNode = document.createElement("select");
+    selectCliLanguageNode.className += "btn text-dark bg-light mx-4";
```
Comment (Ryan-DowlingSoka) [inline source/cli-translate.js:6]: I chose to match the existing class of the <pre> tag for aesthetic reasons. This could be changed to pull the classes from the <pre> instead if necessary.

File: source/cli-translate.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    for(let option of selectCliLanguageOptions){
+        var optionNode = document.createElement("option");
+        optionNode.text=option;
+        selectCliLanguageNode.appendChild(optionNode);
+    }
+    node.after(selectCliLanguageNode);
+    return selectCliLanguageNode;
+}
+
+const imagemagick_linux_to_windows_replacements = new Map([
```
Comment (Ryan-DowlingSoka) [inline source/cli-translate.js:27]: There are two separate maps of replacements. One for Windows command line, the other for additional windows bat files.

File: source/cli-translate.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+function addLinuxWindowsCLIToggle(node) {
+    var copyNode = node.cloneNode(true);
+    copyNode.classList.add("cli-windows");
+    copyNode.style.display = "none";
+    translateLinuxToWindows(copyNode);
+    var batchNode = copyNode.cloneNode(true);
+    batchNode.classList.add("cli-batch");
+    batchNode.style.display = "none";
+    translateWindowsToBatch(batchNode);
+    if(batchNode.innerHTML == node.innerHTML) // No change, ignore.
```
Comment (Ryan-DowlingSoka) [inline source/cli-translate.js:61]: We only have to check the batch file to the linux file because the batch file first runs all windows operations.

File: source/cli-translate.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        return [];
+    }
+
+    node.classList.add("cli-linux");
+    node.after(copyNode);
+    copyNode.after(batchNode);
+    return [copyNode,batchNode];
+}
+
+function addCliTranslations() {
```
Comment (Ryan-DowlingSoka) [inline source/cli-translate.js:73]: This is done in one synchronous block. I did not notice any loading speed issues with this approach. It would be possible to redesign this system to do so more lazily, or over a series of frames, but I don't personally think it is necessary. (it might be more necessary in the imagemagick.org/usage examples since there are many many more examples per page there.)

**Comment by dlemstra** [comment]
Thanks for your contribution. I like this idea but I haven't had time yet to properly test/review this. Can you resolve the conflicts in this pull request?

**Comment by Ryan-DowlingSoka** [comment]
Sure thing. Done.

File: assets/magick-js.php (hunk 1 above)
Comment (dlemstra) [reply assets/magick-js.php:10]: I think I would rather have the `sourceMappingURL` removed from those bootstrap files.

File: source/cli-translate.js (hunk 2 above)
Comment (dlemstra) [inline source/cli-translate.js:1]: Can you move this file to the `assets` folder instead?

File: source/cli-translate.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
+var imagemagick_cli_lists = null;
+
+function addSelectOptions(node)
+{
+    var selectCliLanguageNode = document.createElement("select");
+    selectCliLanguageNode.className += "btn text-dark bg-light mx-4";
+    selectCliLanguageNode.style.fontSize=".75rem";
```
Comment (dlemstra) [inline source/cli-translate.js:7]: Maybe better to have all these style changes as a class inside `magick-template.css`?

File: source/cli-translate.js (hunk 4 above)
Comment (dlemstra) [reply source/cli-translate.js:27]: Sounds like a good plan to do this.

File: source/cli-translate.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    selectCliLanguageNode.style.float="right";
+    selectCliLanguageNode.style.clear="both";
+    selectCliLanguageNode.style.position="relative";
+    selectCliLanguageNode.style.top="-18px";
+    selectCliLanguageNode.style.height="22px";
+    selectCliLanguageNode.style.padding="0px";
+    selectCliLanguageNode.onchange = function(){setCliChanged(this)};
+    selectCliLanguageNode.title = "Change Operating System Syntax";
+    var selectCliLanguageOptions = ["Linux","Windows","Bat File"];
```
Comment (dlemstra) [inline source/cli-translate.js:16]: Maybe `Batch file` instead of `Bat File`?

File: source/cli-translate.js (hunk 6 above)
Comment (dlemstra) [reply source/cli-translate.js:73]: I also tested this locally and this seems to run okay. We can always add a more lazy approach in the future.

File: source/cli-translate.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+function addCliTranslations() {
+    var pres = document.querySelectorAll('pre.cli');
+    var cliWindowsList = [];
+    var cliBatchList = [];
+    var cliLinuxList = [];
+    var cliSelectList = [];
+    for (let pre of pres) {
+        var samp = pre.childNodes[0];
```
Comment (dlemstra) [inline source/cli-translate.js:80]: Could you not do this with a query selector instead?

**Review by dlemstra — COMMENTED** [review]
Thanks for making these changes. Left some review comments.

File: source/cli-translate.js (hunk 9 above)
Comment (Ryan-DowlingSoka) [reply source/cli-translate.js:80]: I probably could. I need the pre and the samp child though, because I addSelectOptions(pre) and I modify the samp, so I'm not sure there is any value in selecting the samp child with the query selector.

Maybe with the query selector I could select only pre.cli that have a samp child, to get a smaller list to iterate over.

File: source/cli-translate.js (hunk 2 above)
Comment (Ryan-DowlingSoka) [reply source/cli-translate.js:1]: Sure.

**Comment by Ryan-DowlingSoka** [comment]
Made the suggested changes. I left two comments on changes, 1 where I'm not sure what you meant by the comment, and the other where I didn't think using a query selector would be beneficial, but am open to being convinced.

File: source/cli-translate.js (hunk 9 above)
Comment (Ryan-DowlingSoka) [reply source/cli-translate.js]: Actually, I guess I could do:
```
var pres = document.querySelectorAll('pre.cli');
for( let pre of pres){
    var samp = pre.querySelector("samp");
    if( samp != null ) {
```

Which may have been what you meant. 👍

**Comment by Ryan-DowlingSoka** [comment]
Updated to a query selector for the samp child.

File: assets/magick-js.php (hunk 1 above)
Comment (Ryan-DowlingSoka) [reply assets/magick-js.php:10]: Ah, I see now. 

//# sourceMappingURL=bootstrap.bundle.min.js.map

**Comment by Ryan-DowlingSoka** [comment]
Removed sourcemappingurl from the bootstrap min.js and remove eol echo into the magick js php.

File: source/cli-translate.js (hunk 9 above)
Comment (dlemstra) [reply source/cli-translate.js]: Yup that is what I meant.

File: source/cli-translate.js (hunk 7 above)
Comment (dlemstra) [reply source/cli-translate.js]: Why did you resolve this conversation?

**Comment by dlemstra** [comment]
Looks like there are still some merging issue with this PR?

File: source/cli-translate.js (hunk 7 above)
Comment (dlemstra) [reply source/cli-translate.js]: Sorry missed this due to merges

**Comment by Ryan-DowlingSoka** [comment]
Interesting it says there arent any merge conflicts here. ill try to merge locally again tonight.

**Comment by dlemstra** [comment]
Thanks, your local `main` branch might be older than you think it is?

**Comment by dlemstra** [comment]
Feel free to reach out by email if you need any help.

**Comment by Ryan-DowlingSoka** [comment]
I hadn't had any time to look at this, but I don't see any conflicts:

image

**Comment by Ryan-DowlingSoka** [comment]
I've updated and merged, without issue, and see no conflicts still. Not sure what to look at next.

**Comment by dlemstra** [comment]
My bad I was seeing this:

afbeelding

But of course this should be squashed instead.

File: include/defines.php [hunk 10]
```diff
@@ -1,4 +1,5 @@
 <div class="magick-header">
+<div id="cli-examples"></div>
```
Comment (dlemstra) [inline include/defines.php:2]: Why was this added?

File: include/import.php [hunk 11]
```diff
@@ -1,4 +1,5 @@
 <div class="magick-header">
+<div id="cli-examples"></div>
```
Comment (dlemstra) [inline include/import.php:2]: And this?

File: include/opencl.php [hunk 12]
```diff
@@ -34,7 +34,7 @@ function
...
-<pre class="bg-light text-dark mx-4"><samp>MAGICK_OCL_DEVICE=true magick sample.jpg -blur 0x5 null:</samp></pre>
+<pre class="bg-light text-dark mx-4 cli"><samp>magick_OCL_DEVICE=true magick sample.jpg -blur 0x5 null:</samp></pre>
```
Comment (dlemstra) [inline include/opencl.php:37]: This should remain `MAGICK_OCL_DEVICE`.

File: include/porting.php [hunk 13]
```diff
@@ -274,7 +274,7 @@
 <h2><a class="anchor" id="deprecate"></a>Deprecated Features Removed</h2>
 <p>All deprecated features from ImageMagick version 6 are removed in version 7.  These include the <samp>Magick-config</samp> and <samp>Wand-config</samp> configuration utilities.  Instead use:</p>
 
-<pre class="bg-light text-dark mx-4"><samp>MagickCore-config
+<pre class="bg-light text-dark mx-4 cli"><samp>magickCore-config
```
Comment (dlemstra) [inline include/porting.php:277]: Can you also revert this?

**Review by dlemstra — COMMENTED** [review]
Found some minor other things. Almost ready to merge.

File: include/defines.php (hunk 10 above)
Comment (Ryan-DowlingSoka) [reply include/defines.php:2]: Ah, this and the other one was missed from an earlier attempt I was making for marking pages that should get the cli treatment. I'll remove these.

**Review by dlemstra — APPROVED** [review]
Thanks for adding this!

**Comment by dlemstra** [comment]
I made some tweaks to your script.  I tried to put this in small commits so it will be easier for you to see what I did.
